core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -603,6 +605,38 @@ public String regexpReplace(String s, String regex, String replacement,
       return Unsafe.regexpReplace(s, pattern, replacement, pos, occurrence);
     }
 
+    /** SQL {@code REGEXP_REPLACE} function with 3 arguments with
+     * {@code \\} based indexing for capturing groups.
+     */
+    public String regexpReplaceNonDollarIndexed(String s, String regex,
+        String replacement) {
+      // Preprocessing to convert double-backslash based indexing for capturing
+      // groups into $ based indices recognized by java regex.
+
+      // Explicitly escaping any $ symbols coming from input
+      // to ignore them from being considered as capturing group index.
+      String indexedReplacement = replacement.replace("\\\\", "\\")
+          .replace("$", "\\$");
+
+      // Check each occurrence of escaped chars, convert \<n> integers into
+      // $<n> indices, keep \\ and \$, throw an error for any other invalid escapes.
+      int lastOccIdx = 0;
+      while (lastOccIdx != -1) {
```

Comment:
I agree and would love to reduce the complexity here, but we need to check every occurrence of double-backslashes left to right to identify if they are used for indexing or escaping other chars ahead. The replacement to $ can only be made if subsequent escaped char is an integer, instead of blindly replacing all double-backslashes.
We can go over it and check if there is an easier path but this is the logic that aligns with how BQ evaluates input expressions, especially in cases where we have multiple backslashes together (example [test case](https://github.com/apache/calcite/blob/166efa4152cd448b9618aba39944215955f2d10f/testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java#L4828))

Reply:
Thanks for the feedback Tanner, restructured the index lookup statement outside the loop to remove one if condition.

Reply:
The code to convert $-format to \-format can be a static utility method.

That utility method can be tested directly in `SqlFunctionsTest`.

You should add a (new) cache in `SqlRegexFunction`. Its type will be a `LoadingCache<String, String>`. Don't worry that most instances of `SqlRegexFunction` won't use this cache. The overhead of creating the cache is small.